### ISSUE TITLE
Aptly publish: fork/exec /usr/bin/bzip2: cannot allocate memory

### ISSUE THREAD

**Issue by karras** [smira/aptly#415]
I'm running an aptly (0.9.7) publish command on a VM (RHEL 6.6) with 4GiB of memory:

`aptly publish -origin=Ubuntu -component=main,restricted,universe,multiverse -distribution=trusty-updates -architectures=amd64,i386,source snapshot trusty-updates-main-stable-20160611T0000Z trusty-updates-restricted-stable-20160611T0000Z trusty-updates-universe-stable-20160611T0000Z trusty-updates-multiverse-stable-20160611T0000Z ubuntu/next-stable`

This will result in the following error:

```
Loading packages...
Generating metadata files and linking package files...
447 / 19076 [====>------------------------------] 2.34 % 2m47
Finalizing metadata files...
ERROR: unable to publish: unable to compress index file: fork/exec /usr/bin/bzip2: cannot allocate memory
```

The aptly process fills more or less the whole memory but doesn't start swapping, beside having 2GiB of swap available.

Why does it not start swapping? Could this be some memory leak / bug?

Thanks!

**Comment by karras** [comment]
The issue doesn't occur with aptly 0.9.5. Seems like this was introduced in 0.9.6 because that version doesn't work either.

Maybe related to #338...

**Comment by demasiadovivo** [comment]
I'm having the same issue after upgrading to 0.9.7. The same repository was working fine when using 0.9.5, but it's failing to publish now. After a couple of seconds aptly consumes all available memory until it crash with the message "ERROR: unable to publish: unable to compress index file: fork/exec /bin/bzip2: cannot allocate memory".
I'm running aptly on ubuntu trusty in a dedicated machine with 2GB of RAM.

**Comment by erickeller** [comment]
I also had a look at the code responsible for generating bz2 contents file seems the following commits are in the suspect list:

```
cave@[~/workspace/go/src/github.com/smira/aptly]$ git tag
 --contains=138f9f7994c2b6fc1470ddd837d709d864a03216
 v0.9.6
 v0.9.7

cave@[~/workspace/go/src/github.com/smira/aptly]$ git tag  --contains=38f4fc209b8ffa6256d475dd2cfaa0c068ae0254
 v0.9.6
 v0.9.7
```

Best Regards

**Comment by sliverc** [comment]
It seems when running publish with flag -skip-contents=true the memory leak doesn't occur

**Comment by sliverc** [comment]
Some more details on this issue. The algorithm to create a repository contents file is using a huge amount of memory - the contents index file feature has been added in version 0.9.6. The contents file is optional as per https://wiki.debian.org/RepositoryFormat#A.22Contents.22_indices but aptly creates it per default (can be disabled with -skip-contents=true)

Format of this file:
FILE LOCATION

FILE is a file system path and LOCATION is the qualified name of a package which installs this file.

Current aptly algorithm checks package by package and checks what files a package will install. To create a contents file format it now keeps those paths in memory as there might be another package referencing the same file.
Only when all packages are checked can then the contents file be written. Keeping this structure in memory causes the memory allocation issues especially with larger repos.

The original post of @karras mentioned following error message:
> ERROR: unable to publish: unable to compress index file: fork/exec /usr/bin/bzip2: cannot allocate memory

When this error message occurs the process has almost made it through the process of writing contents file but failed to compress it. As this compression is actually not needed this is fairly easy to fix and a pull request https://github.com/smira/aptly/pull/449 solves this.

However this is not the only memory allocation issue, there might be other message such like the following:

> panic: unable to unxz data.tar.xz from /aptly/ubuntu/.aptly/pool/82/5e/plasma-widget-redshift_0.6.1-0ubuntu1_amd64.deb: fork/exec /usr/bin/xz: cannot allocate memory

I am currently looking at reducing the memory footprint  to tackle above issue as well.

This just for now as an update.

**Comment by sliverc** [comment]
This memory allocation issue can even occur when enough memory is available.

It seems there is a golang issue where on some systems a fork process cannot allocate memory properly even if there is enough.
See https://github.com/golang/go/issues/5838

As a workaround till golang has fixed this issue can following option being set:
`sysctl -w vm.overcommit_memory=1`

I have tested this with 4GB memory and the publish process for trusty finished successfully (2GB was not enough and it failed).

**Comment by sliverc** [comment]
Finally I have rewritten the contents index algorithm to solve this problem.

See https://github.com/smira/aptly/pull/454

**Comment by smira** [comment]
Should be fixed with #490
